test-data/unit/check-expressions.test

```diff
@@ -300,6 +300,28 @@ b = a and b # E: Incompatible types in assignment (expression has type "Union[A,
 b = b or a  # E: Incompatible types in assignment (expression has type "Union[bool, A]", variable has type "bool")
 b = a or b  # E: Incompatible types in assignment (expression has type "Union[A, bool]", variable has type "bool")
 class A: pass
+
+[builtins fixtures/bool.py]
+
+[case testRestrictedTypeAnd]
+
+b = None # type: bool
+i = None # type: str
+j = not b and i
+if j:
+    s = j # type: str
```

Comment:
Minor nit: if you want to make sure a variable is a certain type, it's a bit better to use `reveal_type` because otherwise it could be `Any` (or the branch might not be checked).